restful.entity.inc

```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
+    'bundles' => array(),
+    'bundle keys' => array(
+      'bundle' => 'type',
+    ),
+    'module' => 'restful',
+    'entity cache' => FALSE,
+  );
 
   return $items;
 }
 
 /**
+ * Helper function that extract cache hashes from an entity.
+ */
+function _restful_entity_cache_hashes($entity, $type) {
+  if ($type == 'cache_fragment') {
+    return array();
+  }
+  // Limit to the fragments for our entity type.
+  $query = new \EntityFieldQuery();
+  $query
+    ->entityCondition('entity_type', 'cache_fragment')
+    ->propertyCondition('type', 'entity_type')
+    ->propertyCondition('value', $type);
+  if (!$hashes = CacheFragmentController::lookUpHashes($query)) {
+    return array();
+  }
+
+  // Search for all the cache fragments with our entity id.
+  list($entity_id,,) = entity_extract_ids($type, $entity);
+  $query = new \EntityFieldQuery();
+  $query
+    ->entityCondition('entity_type', 'cache_fragment')
+    ->propertyCondition('type', 'entity_id')
+    ->propertyCondition('value', $entity_id)
+    ->propertyCondition('hash', $hashes, 'IN');
+  return CacheFragmentController::lookUpHashes($query);
+}
+
+/**
  * Implements hook_entity_update().
  */
 function restful_entity_update($entity, $type) {
-  list($entity_id) = entity_extract_ids($type, $entity);
-  $cid = 'paet:' . $type . '::ei:' . $entity_id;
-  CacheDecoratedDataProvider::invalidateEntityCache($cid);
+  $resource_manager = restful()->getResourceManager();
+  foreach (_restful_entity_cache_hashes($entity, $type) as $hash) {
+    $handler = $resource_manager->getPlugin(CacheFragmentController::resourceIdFromHash($hash));
+    if (!$handler instanceof CacheDecoratedResourceInterface) {
+      return;
+    }
+    if (!$handler->hasSimpleInvalidation()) {
+      return;
+    }
+    $handler
+      ->getCacheController()
+      ->clear($hash);
+  }
 }
 
 /**
  * Implements hook_entity_delete().
  */
 function restful_entity_delete($entity, $type) {
-  list($entity_id) = entity_extract_ids($type, $entity);
-  $cid = 'paet:' . $type . '::ei:' . $entity_id;
-  CacheDecoratedDataProvider::invalidateEntityCache($cid);
+  $resource_manager = restful()->getResourceManager();
+  foreach (_restful_entity_cache_hashes($entity, $type) as $hash) {
+    $handler = $resource_manager->getPlugin(CacheFragmentController::resourceIdFromHash($hash));
+    if (!$handler instanceof CacheDecoratedResourceInterface) {
+      return;
+    }
+    if (!$handler->hasSimpleInvalidation()) {
+      return;
+    }
+    $handler
+      ->getCacheController()
+      ->clear($hash);
+  }
 }
 
 /**
  * Implements hook_user_update().
  */
 function restful_user_update(&$edit, $account, $category) {
-  // Due to the limitations for cid matching on clearing caches, we need to
-  // clear all the bin. We cannot do $cid = '%::uu' . $account->uid . '::pa';
-  CacheDecoratedDataProvider::invalidateEntityCache('*');
+  // Search for all the cache fragments with our entity id.
+  $query = new \EntityFieldQuery();
```

Comment:
I think It's because of the granularity of the cache clear - https://github.com/RESTful-Drupal/restful/blob/f0f981a8552a8c8b429bce2991d06e551661adff/restful.entity.inc#L56-L58